### PR TITLE
Refactor to allow overriding type scale

### PR DESCRIPTION
**Pull request by james-lee-ck**
This PR adds the ability to override the typescale. This allows us to configure the sizes for the various font sizes, as well as give us correct sizes in the `sizes` object.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ const hyphensToUnderscores = function(sourceObj) {
 
     /* Create hypened versions */
     _.forOwn(sourceObj, (val, key) => {
-        translated[key.replace(/-/g, "_")] = val;
+        translated[key.replace(/-/ug, "_")] = val;
```
Comment (james-lee-ck) [inline src/index.js]: 💁 Unrelated, but was failing linting

File: src/reactWrapper.js [hunk 2]
```diff
@@ -47,7 +46,7 @@ function setStyles(props, clsPropName) {
         newProps.style = []
     }
 
-    const splitted = props[clsPropName].replace(/-/g, "_").split(" ")
+    const splitted = props[clsPropName].replace(/-/ug, "_").split(" ")
```
Comment (james-lee-ck) [inline src/reactWrapper.js]: 💁 Unrelated, but failing linting here, on L:68 and L:86

File: src/index.js [hunk 3]
```diff
@@ -60,6 +61,8 @@ const NativeTachyons = {
         _.assign(styleSheet, text)
         _.assign(styleSheet, opacity)
 
+        const typeScale = generateTypeScale(options.typeScale)
```
Comment (james-lee-ck) [inline src/index.js]: 💁 Generates the typescale using the merged `defaultTypeScale` and passed in `typeScale` configuration.

**Comment by james-lee-ck** [comment]
Could I get some :eyes: on this please? Hoping to get this (or something like it) for better configuration of font sizes. I'm happy to iterate on changes.

**Comment by fab1an** [comment]
Why do you need to modify the type-scale? Do you need additional steps, or just a linear scaling?

**Comment by james-lee-ck** [comment]
Both actually. And it's not necessarily pure linear scaling. We have an existing styleguide that we are following for both our mobile and web app, and the default scaling doesn't work well for us in the mobile scenario.

**Comment by fab1an** [comment]
Could you write down the typescale you would use here, even better with rem settings for dpi?

**Comment by james-lee-ck** [comment]
The typescale in the documentation I wrote is pretty close to the typescale I would use.

**Comment by james-lee-ck** [comment]
Actually, its slightly tweaked, it would be:

```
    f1: '1.625',
    f2: '1.375',
    f3: '1.125',
    f4: '1.0',
    f5: '0.875',
    f6: '0.75',
    f7: '0.625',
```

**Comment by james-lee-ck** [comment]
Could you maybe explain what your reservations are? From my perspective, since I'm integrating into an existing application that has an established set of font sizes and styleguide, the default scaling won't be possible. The default scaling scales too rapidly to be usable for my mobile only app. 

This PR just adds the option to configure the scaling to allow more fine tuned sizes, while still falling back to the defaults if not provided.

**Comment by fab1an** [comment]
Assuming that you have read or skimmed (http://mrmrs.cc/writing/2016/03/24/scalable-css/ and http://tachyons.io/docs/typography/scale/).

My perspective is that I want to keep rnst as simple as possible and as close to the original tachyons.css as possible. Tachyons' benefit apart from other things is that it enables non-designer to create pretty layouts, by purposely restricting what you can do.

At the same time I know that the original classes don't always make sense, especially the largest type-settings are much more useful in print and maybe on websites than anywhere mobile.

So whenever a user/developer asks for modification of the default scales, I try to understand why it is necessary, and why no other developer had this necessity before.

Also, since the entire scale is multiplied by REM, it's possible that someone started with a "wrong" rem-setting and the need for another step in the scale is actually remedied by modifying the rem and maybe shifting all layouts one step on the scale (f2->f3, f3->f4, ...).

Last but not least, there in an unsolved issue which should be fixed, the line-height calculation is wrong #65

**Comment by fab1an** [comment]
I agree. I think it’s easy with rnst as well, you can just pass configure your typescale in an object and add it to styles.
But granted, I only had to do this in a couple of situations, not with every <Text>. 
Do you need to override the spacing scale as well?

**Comment by fab1an** [comment]
I will probably go with your proposal, thank you for the explanations. If you want it you can share a sample-screenshot of your design.

**Comment by james-lee-ck** [comment]
Regarding configuring by passing an object, thats actually a route that we initially went down. But since we sometimes use the `sizes` object to do calculations, we ended up having to override that as well, which felt a little dirty. 

Here's an example of our original overrides:

```
export const typeScales = {
    f1: 1.625,
    f2: 1.375,
    f3: 1.125,
    f4: 0.9375,
    f5: 0.8125,
    f6: 0.75,
    f7: 0.625,
}

const typeSizes = Object.keys(typeScales).reduce((acc, name) => {
    acc[name] = typeScales[name] * rem
    return acc
}, {})

const customStyles = Object.keys(typeScales).reduce((acc, name) => {
    acc[name] = { fontSize: typeSizes[name] }
    return acc
}, {})

NativeTachyons.build(
    {
        colors: {
            palette: colors,
        },
        customStyles,
        fonts,
        rem,
    },
    StyleSheet
)

const { sizes } = NativeTachyons
const sizesWithTypeSizes = Object.assign({}, sizes, typeSizes)

export const { wrap, options } = NativeTachyons
export { sizesWithTypeSizes as sizes }
```

**Comment by james-lee-ck** [comment]
Is there anything I can help with to get this merged? If there are changes you'd like to see, I'd be happy to refactor.

**Comment by fab1an** [comment]
Are you still interesting in working on this, even though it was dormant for over a year?

**Comment by fab1an** [comment]
Merged, thank you.

**Pull request merged**